runtime/common/src/parachains.rs

```diff
@@ -142,6 +231,145 @@ pub enum Origin {
+#[derive(RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct DoubleVoteOffence<Offender> {
+	/// The current session index in which we report a validator.
+	session_index: SessionIndex,
+	/// The size of the validator set in current session/era.
+	validator_set_count: u32,
+	/// An offender that has submitted two conflicting votes.
+	offender: Offender,
+}
+
+impl<Offender: Clone> Offence<Offender> for DoubleVoteOffence<Offender> {
+	const ID: Kind = *b"double-vote:doub";
+	type TimeSlot = SessionIndex;
+
+	fn offenders(&self) -> Vec<Offender> {
+		vec![self.offender.clone()]
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.session_index
+	}
+
+	fn validator_set_count(&self) -> u32 {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.session_index
+	}
+
+	fn slash_fraction(_offenders_count: u32, _validator_set_count: u32) -> Perbill {
+		// Slash 100%.
+		Perbill::from_percent(100)
+	}
+}
+
+impl<O: Clone> ParachainsOffence<O> for DoubleVoteOffence<O> {
+	fn new(
+		session_index: SessionIndex,
+		validator_set_count: u32,
+		offender: O,
+	) -> Self {
+		Self {
+			session_index,
+			validator_set_count,
+			offender,
+		}
+	}
+}
+
+pub trait ParachainsOffence<O>: Offence<O> {
+	fn new(
+		session_index: SessionIndex,
+		validator_set_count: u32,
+		offender: O,
+	) -> Self;
+}
+
+pub trait HandleDoubleVote<T: Trait> {
+	/// The proof of key ownership.
+	type Offence: ParachainsOffence<IdentificationTuple<T>>;
+
+	fn check_report(
+		report: &DoubleVoteReport,
+		key_owner_proof: Proof,
+		i: Vec<u8>
+	) -> Option<IdentificationTuple<T>>;
+
+	fn report_offence(
+		reporters: Vec<T::AccountId>,
+		offence: Self::Offence
+	);
+}
+
+impl<T: Trait> HandleDoubleVote<T> for () {
+	type Offence = DoubleVoteOffence<IdentificationTuple<T>>;
+
+	fn check_report(
+		_report: &DoubleVoteReport,
+		_key_owner_proof: Proof,
+		_i: Vec<u8>,
+	) -> Option<IdentificationTuple<T>> {
+		None
+	}
+
+	// TODO: As soon as substrate is updated and `report_offence` returns
+	// a result, this function should also be changed to returning a result.
+	fn report_offence(
+		_reporters: Vec<T::AccountId>,
+		_offence: Self::Offence,
+	) {}
+}
+
+pub struct DoubleVoteHandler<P, R, O> {
+	_phantom: rstd::marker::PhantomData<(P, R, O)>,
+}
+
+impl<P, R, O> Default for DoubleVoteHandler<P, R, O> {
+	fn default() -> Self {
+		Self {
+			_phantom: Default::default(),
+		}
+	}
+}
+
+impl<T, P, R, O> HandleDoubleVote<T> for DoubleVoteHandler<P, R, O>
+where
+	T: Trait<HandleDoubleVote = Self>,
+	P: KeyOwnerProofSystem<
+		// AccountId?
+		(sp_core::crypto::KeyTypeId, Vec<u8>),
+		Proof = Proof,
+		IdentificationTuple = session::historical::IdentificationTuple<T>,
+	>,
+	O: ParachainsOffence<P::IdentificationTuple>,
+	R: ReportOffence<T::AccountId, P::IdentificationTuple, O>,
+{
+	type Offence = O;
+
+	fn check_report(
+		_report: &DoubleVoteReport,
+		key_owner_proof: Proof,
+		i: Vec<u8>,
```

Comment:
In light of my previous comment, this parameter probably doesn't make sense as the `ValidatorId` is extractable from the `report` parameter.